### PR TITLE
[OpenMP] [amdgpu] Added a synchronous version of data exchange.

### PR DESCRIPTION
**Pull request by dhruvachak**
Similar to H2D and D2H, use synchronous mode for large data transfers
beyond a certain size for D2D as well. As with H2D and D2H, this size is
controlled by an env-var.

### CODE REVIEW

**Comment by jhuber6** [comment]
Can you update the description to describe why we want synchronous data exchange?

**Comment by dhruvachak** [comment]
I amended the commit message to include the following:

Similar to H2D and D2H, use synchronous mode for large data transfers
beyond a certain size for D2D as well. As with H2D and D2H, this size is
controlled by an env-var.

File: openmp/libomptarget/test/offloading/d2d_memcpy_sync.c [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+  const int src_device = 0;
+  int dst_device = num_devices - 1;
+
+  int length = N * sizeof(int);
+  int *src_ptr = omp_target_alloc(length, src_device);
+  int *dst_ptr = omp_target_alloc(length, dst_device);
+
+  assert(src_ptr && "src_ptr is NULL");
+  assert(dst_ptr && "dst_ptr is NULL");
```
Comment (arsenm) [inline openmp/libomptarget/test/offloading/d2d_memcpy_sync.c:31]: don't assert to check allocation failure

File: openmp/libomptarget/test/offloading/d2d_memcpy_sync.c [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+  }
+
+  int rc =
+      omp_target_memcpy(dst_ptr, src_ptr, length, 0, 0, dst_device, src_device);
+
+  assert(rc == 0 && "error in omp_target_memcpy");
+
+  int *buffer = malloc(length);
+
+  assert(buffer && "failed to allocate host buffer");
```
Comment (arsenm) [inline openmp/libomptarget/test/offloading/d2d_memcpy_sync.c:46]: same

**Comment by jhuber6** [comment]
> I amended the commit message to include the following:
> 
> Similar to H2D and D2H, use synchronous mode for large data transfers beyond a certain size for D2D as well. As with H2D and D2H, this size is controlled by an env-var.

You should do it to the PR message as well. When the PR is merged it will squash the commits and use the PR's description instead.

**Comment by dhruvachak** [comment]
> > I amended the commit message to include the following:
> > Similar to H2D and D2H, use synchronous mode for large data transfers beyond a certain size for D2D as well. As with H2D and D2H, this size is controlled by an env-var.
> 
> You should do it to the PR message as well. When the PR is merged it will squash the commits and use the PR's description instead.

Yes, I just realized that the PR comment was not updated. Now it should be updated.

File: openmp/libomptarget/test/offloading/d2d_memcpy_sync.c (hunk 1 above)
Comment (dhruvachak) [reply openmp/libomptarget/test/offloading/d2d_memcpy_sync.c]: Changed to check and FAIL if required. The earlier change was based on an existing test which asserted.

File: openmp/libomptarget/test/offloading/d2d_memcpy_sync.c (hunk 2 above)
Comment (dhruvachak) [reply openmp/libomptarget/test/offloading/d2d_memcpy_sync.c]: done.

**Review by jhuber6 — APPROVED** [review]
LG, thanks

What's the expected behavior when doing a D2D memcpy onto the same device? Just wondering if that's a no-op under the hood for the purposes of the buildbot.

**Comment by dhruvachak** [comment]
> LG, thanks
> 
> What's the expected behavior when doing a D2D memcpy onto the same device? Just wondering if that's a no-op under the hood for the purposes of the buildbot.

The plugin still uses the same async memcpy. I don't know whether there is some underlying optimization for the same-device case.

**Pull request merged**